stdlib/src/bisect.rs

```diff
@@ -94,20 +111,36 @@ mod _bisect {
     #[inline]
     #[pyfunction]
     fn bisect_right(
-        BisectArgs { a, x, lo, hi }: BisectArgs,
+        BisectArgs { a, x, lo, hi, key }: BisectArgs,
         vm: &VirtualMachine,
     ) -> PyResult<usize> {
         let (mut lo, mut hi) = as_usize(lo, hi, a.length(vm)?, vm)?;
 
-        while lo < hi {
-            // Handles issue 13496.
-            let mid = (lo + hi) / 2;
-            if x.rich_compare_bool(&*a.get_item(mid, vm)?, Lt, vm)? {
-                hi = mid;
-            } else {
-                lo = mid + 1;
+        if let Some(ref key) = key {
+            while lo < hi {
+                let mid = (lo + hi) / 2;
+                let a_mid = a.get_item(mid, vm)?;
+                let comp = vm.invoke(key, (a_mid.clone(),))?;
+                if x.rich_compare_bool(&comp, Lt, vm)? {
+                    hi = mid;
+                } else {
+                    lo = mid + 1;
+                }
+            }
+
+        } else {
+            while lo < hi {
+                // Handles issue 13496.
+                let mid = (lo + hi) / 2;
+                if x.rich_compare_bool(&*a.get_item(mid, vm)?, Lt, vm)? {
+                    hi = mid;
+                } else {
+                    lo = mid + 1;
+                }
```

Comment:
this part can be merged with `Some(key)` part. 
they share everything except for key invoke.

none checking is very cheap comparing to `get_item`.